proposals/NNNN-isolated-synchronous-deinit.md

```diff
@@ -0,0 +1,386 @@
+
+### Improving de-virtualization and inlining of the executor access.
+
+Consider the following example:
+
+```swift
+import Foundation
+
+@_silgen_name("do_it")
+@MainActor func doIt()
+
+public class Foo {
+  @MainActor
+  public func foo() async {
+    doIt()
+  }
+  @MainActor
+  deinit {}
+}
+```
+
+Currently both method and `deinit` entry points produce two calls to access the `MainActor.shared.unownedExecutor`, with the second one even using dynamic dispatch.
+These two calls could be replaced with a single call to the statically referenced `swift_task_getMainExecutor()`.
+
+```llvm
+%1 = tail call swiftcc %swift.metadata_response @"type metadata accessor for Swift.MainActor"(i64 0) #6
+%2 = extractvalue %swift.metadata_response %1, 0
+%3 = tail call swiftcc %TScM* @"static Swift.MainActor.shared.getter : Swift.MainActor"(%swift.type* swiftself %2)
+%4 = tail call i8** @"lazy protocol witness table accessor for type Swift.MainActor and conformance Swift.MainActor : Swift.Actor in Swift"() #6
+%5 = bitcast %TScM* %3 to %objc_object*
+%6 = tail call swiftcc { i64, i64 } @"dispatch thunk of Swift.Actor.unownedExecutor.getter : Swift.UnownedSerialExecutor"(%objc_object* swiftself %5, %swift.type* %2, i8** %4)
+```
+
+### Making fast path inlinable
+
+For this to be useful compiler first needs to be able to reason about the value of the current executor based on the isolation of the surrounding function. Then inlinable pre-check for `swift_task_isCurrentExecutor()` can be inserted allowing isolated deallocating deinit to be inlined into non-isolated one.
+
+### Improving extended stack trace support
+
+Developers who put breakpoint in the isolated deinit might want to see the call stack that lead to the least release of the object. Currently if switching of executors was involved, release call stack won't be shown in the debugger.
+
+### Implementing API for synchronously scheduling arbitrary work on the actor
+
+Introduced runtime functions has calling convention optimized for deinit use case, but using similar runtime function with a slightly different signature, one could implement an API for synchronously scheduling arbitrary work on the actor:
+
+```swift
+extension Actor {
+  /// Adds a job to the actor queue that calls `work` passing `self` as an argument.
+  nonisolated func enqueue(_ work: __owned @Sendable @escaping (isolated Self) -> Void)
+
+  /// If actor's executor is already the current one - executes work immediately
+  /// Otherwise adds a job to the actor's queue.
+  nonisolated func executeOrEnqueue(_ work: __owned @Sendable @escaping (isolated Self) -> Void)
+}
+
+actor MyActor {
+    var k: Int = 0
+    func inc() { k += 1 }
+}
+
+let a = MyActor()
+a.enqueue { aIsolated in
+    aIsolated.inc() // no await
+}
+```
+
+## Alternatives considered
```

Comment:
I think Joe's comment is complementary to this proposal, not really an alternative. I describes how move-only types could have isolated deinit which does not have isolation thunk, because deinit is called semi-explicitly (using `drop()` function or similar tool) and developers need to manually ensure that object is destroyed in correct isolation context. While this proposal focuses on supporting types with shared ownership.

I think I can incorporate a paragraph about move-only types inspired by Joe's comment into future directions section instead.